docs/RFCS/change_feeds_primitive.md

```diff
@@ -0,0 +1,362 @@
+should be straightforward to add an entity which abstracts away the Range level.
+For example, given the key span of a whole table, that entity splits it along
+range boundaries, and opens individual `ChangeFeed` streams to suitable members
+of each range. When individual streams disconnect (for instance due to a split)
+new streams are initiated (using the last known closed timestamp for the lost
+stream).
+
+## Optimization: resume tokens
+
+To avoid unnecessary catch-up on already received values, a straightforward
+optimization is possible for the common case in which a stream disconnects
+because of a replication or lease change (but not a split). Namely,
+
+1. each `Value()` event contains a `resume_token`, which is opaque to the caller
+   but totally ordered. It is really a Raft log index of the Raft log entry
+   that has triggered the event.
+1. when a stream drops, the caller reconnects with both the highest known
+   checkpointed timestamp *and* the largest seen `resume_token` (if any).
+1. if the Replica serving the new stream has the log index corresponding to
+   `resume_token+1` available (i.e. if it hasn't been truncated away), it skips
+   the catch-up phase by sending `StreamState` first, and then replays events
+   straight from the Raft log. Otherwise, it behaves as if no `resume_token` was
+   specified.
+1. Whenever the caller receives `StreamState` as the first event on a
+   reconnected stream, it knows that it won't be receiving any duplicate values
+   (callers can either deduplicate by keeping state or drop intermediate results
+   on reconnection; when `resume_token` catch up works frequently, it is better
+   to do the latter). If instead it receives an event, it needs to be prepared
+   to handle duplicates.
+1. Since log-based catch-up is more effective, `resume_token` should be
+   specified whenever it is available, even if the caller already keeps state
+   for deduplication.
+
+## Implementation suggestions
+
+There is a straightforward path to implement small bits and pieces of this
+functionality without embarking on an overwhelming endeavour all at once:
+
+First, implement the basic `ChangeFeed` command, but without the base timestamp
+or checkpoints. That is, once registered with `Raft`, updates are streamed, but
+there is no information on which updates were missed and what timestamps to
+expect. In turn, the MVCC work is not necessary yet, and neither are [follower
+reads][followerreads].
+
+Next, add an experimental `./cockroach debug changefeed <key>` command which
+launches a single `ChangeFeed` request through `DistSender`. It immediately
+prints results to the command line without buffering and simply retries the
+`ChangeFeed` command until the client disconnects. It may thus both produce
+duplicates and miss updates.
+
+This is a toy implementation that already makes for a good demo and allows a
+similar toy implementation that uses [streaming SQL][streamsql] once it's
```

Comment:
Need to clarify somewhere what this "streaming SQL" feature is that you're referring to multiple times.